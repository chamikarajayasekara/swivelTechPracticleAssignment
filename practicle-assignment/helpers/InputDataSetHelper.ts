var validator = require('validator');

export function InitiallySetData(fields:any, action:string, setInputValue:any, setInputErrorValue:any, selectedEmployee:any, setEditFields:any) {
    for (let i = 0; i < fields.length; i++) {
        let relatedField = fields[i]
        setEditFields((prev:any) => ({
            ...prev,
            [relatedField['dbName']]: false,
        }));
        setInputErrorValue((prev:any) => ({
            ...prev,
            [relatedField['dbName']]: {"error":false, message:""},
        }));
        if (action === "add"){
            setInputValue((prev:any) => ({
                ...prev,
                [relatedField['dbName']]:"",
            }));

        }else{
            setInputValue((prev:any) => ({
                ...prev,
                [relatedField['dbName']]: selectedEmployee[relatedField['dbName']],
            }));
        }
    }
}

export function InputValueGet(inputValue:any, dbName:string) {
    return inputValue[dbName];
}

export function errorIdentifiedInSubmit(fields:any,obj:any,setInputErrorValue:any, setEditFields:any ){
    for (let i = 0; i < fields.length; i++) {
        let relatedField = fields[i];
        for (var key in obj) {
            if (obj[key] === null || obj[key] === ""){
                if (relatedField.dbName === key){
                    let errorMessage: string = relatedField.errorMessage;
                    let errorField = relatedField.dbName;
                    setInputErrorValue((prev:any) => ({
                        ...prev,
                        [errorField ]: {"error":true, "message": errorMessage},
                    }));
                    setEditFields((prev:any) => ({
                        ...prev,
                        [errorField]: true,
                    }));
                }
            }else {
                if (relatedField.dbName === key){
                    let errorMessage: string = relatedField.errorMessage;
                    let errorField = relatedField.dbName;
                    if (key === "firstName"){
                        if(obj[key].length >=  6 && obj[key].length <= 10){
                            setInputErrorValue((prev:any) => ({
                                ...prev,
                                [errorField ]: {"error":false, "message":""},
                            }));
                        }else{
                            setInputErrorValue((prev:any) => ({
                                ...prev,
                                [errorField]: {"error":true, "message":"min 6 character and max 10 characters."},
                            }));
                        }
                    }

                    if (key === "lastName"){
                        if(obj[key].length >=  6 && obj[key].length <= 10){
                            setInputErrorValue((prev:any) => ({
                                ...prev,
                                [errorField ]: {"error":false, "message":""},
                            }));
                        }else{
                            setInputErrorValue((prev:any) => ({
                                ...prev,
                                [errorField]: {"error":true, "message":"min 6 character and max 10 characters."},
                            }));
                        }
                    }
                    if (key === "emailAddress"){
                        if(validator.isEmail(obj[key])){
                            setInputErrorValue((prev:any) => ({
                                ...prev,
                                [errorField ]: {"error":false, "message":""},
                            }));
                        }else{
                            setInputErrorValue((prev:any) => ({
                                ...prev,
                                [errorField]: {"error":true, "message":"please add valid email."},
                            }));
                        }
                    }


                    if (key === "phoneNumber"){
                        const regex = /^(?:0|94|\+94)?(?:(11|21|23|24|25|26|27|31|32|33|34|35|36|37|38|41|45|47|51|52|54|55|57|63|65|66|67|81|912)(0|2|3|4|5|7|9)|7(0|1|2|4|5|6|7|8)\d)\d{6}$/;
                        let matches = regex.test(obj[key]);
                        if(matches){
                            setInputErrorValue((prev:any) => ({
                                ...prev,
                                [errorField ]: {"error":false, "message":""},
                            }));
                        }else{
                            setInputErrorValue((prev:any) => ({
                                ...prev,
                                [errorField]: {"error":true, "message":"Please add valid phone number."},
                            }));
                        }
                    }
                    // setInputErrorValue((prev:any) => ({
                    //     ...prev,
                    //     [errorField ]: {"error":false, "message":""},
                    // }));
                    // setEditFields((prev:any) => ({
                    //     ...prev,
                    //     [errorField]: false,
                    // }));

                }
            }
        }
    }
}


export function ErrorsExists(editFields:any) {
    const areTrue = Object.values(editFields).some(
        value => value === true
    );
    return areTrue;
}


export function checkIsNullishByAll(obj:any, fields:any ) {
    let result = false;
    let resultArray:any = [];
    for (let i = 0; i < fields.length; i++) {
        let relatedField: any = fields[i];
            if(relatedField.mandatory === true){
                if ( relatedField.dbName in obj ){
                    if (obj[relatedField.dbName] === null || obj[relatedField.dbName]  === ""){
                        resultArray.push('true')
                    }else{
                        resultArray.push('false')
                    }
                }
            }else{
                resultArray.push('false')
            }
        }
    resultArray.some((x:string) => {
        if (x === 'true'){
            result = true;
        }
    })
    return result;
}