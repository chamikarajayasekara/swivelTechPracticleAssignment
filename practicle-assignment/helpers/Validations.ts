import * as Yup from "yup";

const phoneRegExp = /^(?:0|94|\+94)?(?:(11|21|23|24|25|26|27|31|32|33|34|35|36|37|38|41|45|47|51|52|54|55|57|63|65|66|67|81|912)(0|2|3|4|5|7|9)|7(0|1|2|4|5|6|7|8)\d)\d{6}$/;

export type UserSubmitForm = {
    firstName: string;
    lastName: string;
    emailAddress: string;
    phoneNumber: string;
    gender: string,
    photo:string
};
export const validationSchema = Yup.object().shape({
    firstName: Yup.string()
        .required('First Name is required')
        .min(6, 'First Name must be at least 6 characters')
        .max(10, 'First Name must not exceed 10 characters'),
    lastName: Yup.string()
        .required('Last Name is required')
        .min(6, 'Last Name must be at least 6 characters')
        .max(10, 'Last Name must not exceed 10 characters'),
    emailAddress: Yup.string()
        .required('Email is required')
        .email('Email is invalid'),
    phoneNumber: Yup.string().matches(phoneRegExp, 'Phone number is not valid'),
    gender: Yup.string()
        .required('gender is required')
});