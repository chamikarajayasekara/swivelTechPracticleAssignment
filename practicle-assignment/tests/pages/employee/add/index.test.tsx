import {fireEvent, render, renderHook, screen} from '@testing-library/react';
import Index from "../../../../pages/employee/add/index";
import Home from "../../../../pages";
import {createMockRouter} from "../../../../helpers/createMockrouter";
import { RouterContext } from 'next/dist/shared/lib/router-context';
const useRouter = jest.spyOn(require('next/router'), 'useRouter')



describe('Home', () => {
    it('renders a description', () => {
        const { container } = render(<Index/>)
        const textToFind = "LIST VIEW"
        const a = screen.getByText(textToFind)
        expect(a).toBeInTheDocument();
        expect(container).toMatchSnapshot()
    })

    it('calls router.push with "/contacts', () => {
        const router = createMockRouter({
            // query: { id: '44' },
            pathname: 'add',
            basePath: 'employee',
        });

        render(
            <RouterContext.Provider value={router}>
                <Index/>;
            </RouterContext.Provider>
        );

        fireEvent.click(screen.getByRole('button'));
        // expect(router.push).toHaveBeenCalledWith(
        //     '/employee/list'
        // );
    });
})
