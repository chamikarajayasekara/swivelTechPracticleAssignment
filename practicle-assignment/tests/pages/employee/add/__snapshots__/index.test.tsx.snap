// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders a description 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col-sm-10"
      />
      <div
        class="col-sm-2"
      >
        <div
          class="grid-select-area"
        >
          <div
            class="roundedButton"
          >
            LIST VIEW
          </div>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-sm-2"
      />
      <div
        class="col-sm-8"
      >
        <div
          class="card"
        >
          <form>
            <div
              class="input-section"
            >
              <div
                class="label-line"
              >
                <label
                  for=""
                >
                  First Name
                </label>
              </div>
              <div
                class="textField"
              >
                <div
                  class="error-line"
                />
                <input
                  class="input-control"
                  name="First Name"
                  type="input"
                  value=""
                />
              </div>
            </div>
            <div
              class="input-section"
            >
              <div
                class="label-line"
              >
                <label
                  for=""
                >
                  Last Name
                </label>
              </div>
              <div
                class="textField"
              >
                <div
                  class="error-line"
                />
                <input
                  class="input-control"
                  name="Last Name"
                  type="input"
                  value=""
                />
              </div>
            </div>
            <div
              class="input-section"
            >
              <div
                class="label-line"
              >
                <label
                  for=""
                >
                  Email
                </label>
              </div>
              <div
                class="textField"
              >
                <div
                  class="error-line"
                />
                <input
                  class="input-control"
                  name="Email"
                  type="input"
                  value=""
                />
              </div>
            </div>
            <div
              class="input-section"
            >
              <div
                class="label-line"
              >
                <label
                  for=""
                >
                  Phone
                </label>
              </div>
              <div
                class="textField"
              >
                <div
                  class="error-line"
                />
                <input
                  class="input-control"
                  name="Phone"
                  type="input"
                  value=""
                />
              </div>
            </div>
            <div
              class="input-section"
            >
              <div
                class="label-line"
              >
                <label
                  for=""
                >
                  Gender
                </label>
              </div>
              <div
                class="textField"
              >
                <div
                  class="error-line"
                />
                <select
                  class="input-control"
                  id="Gender"
                  name="Gender"
                  required=""
                >
                  <option
                    class="option-control"
                    disabled=""
                  >
                    Gender
                  </option>
                  <option
                    class="option-control"
                  >
                    M
                  </option>
                  <option
                    class="option-control"
                  >
                    F
                  </option>
                </select>
              </div>
            </div>
            <div
              class="footer-section"
            >
              <button
                class="borderButton"
              >
                ADD
              </button>
            </div>
          </form>
        </div>
      </div>
      <div
        class="col-sm-2"
      />
    </div>
  </div>
</div>
`;
